subsys/bluetooth/host/adv.c

```diff
@@ -1340,7 +1340,7 @@ int bt_le_adv_start(const struct bt_le_adv_param *param,
 	int err;
 
 	if (!adv) {
-		return -ENOMEM;
+		return -EALREADY;
```

Comment:
I agree with the main commit message, but I think this enomem is valid since we're allocating a context. I think we should rather test the bit earlier when EXT ADV is disabled.
```c
	if (atomic_test_bit(adv->flags, BT_ADV_ENABLED)) {
		return -EALREADY;
	}
```

Reply:
The `adv_new_legacy()` is a bit odd, because when extended adv is enable, it does, as you say, allocate a context (which also matches the name of the function itself). However, when extended advertising is not enabled, then it actually does not allocate a context, but rather returns the existing one. 

Maybe the `adv_new_legacy()` should be refactored to work the same for `CONFIG_BT_EXT_ADV=y` and `CONFIG_BT_EXT_ADV=n`, so that it may also return existing advertiser when `CONFIG_BT_EXT_ADV=y`. 

That would also remove this check (which is only valid for `CONFIG_BT_EXT_ADV=y` btw), and then the calls to `bt_le_adv_start_ext` or `bt_le_adv_start_legacy` would provide the proper state check. 

So perhaps change the function from
```c
static struct bt_le_ext_adv *adv_new_legacy(void)
{
#if defined(CONFIG_BT_EXT_ADV)
	if (bt_dev.adv) {
		return NULL;
	}

	bt_dev.adv = adv_new();
	return bt_dev.adv;
#else
	return &bt_dev.adv;
#endif
}
```
to
```c
static struct bt_le_ext_adv *adv_new_legacy(void)
{
#if defined(CONFIG_BT_EXT_ADV)
	if (bt_dev.adv) {
		return bt_dev.adv;
	}

	bt_dev.adv = adv_new();
	return bt_dev.adv;
#else
	return &bt_dev.adv;
#endif
}
```

and leave this check as is (as `adv_new` may still return NULL. 

That way we get the right return value for both configs in case that it is already started

Reply:
And perhaps `adv_new_legacy` should be renamed `adv_get_or_new_legacy`

Reply:
I second @Thalley's solution of fixing `adv_new_legacy`. I think `adv_get_legacy` or `adv_get_legacy_instance` may be a good new name. It's similar to `get_instance` that is commonly used for singletons.